templates/common/infra/bicep/core/host/container-app.bicep

```diff
@@ -5,24 +5,46 @@ param tags object = {}
 param containerAppsEnvironmentName string
 param containerName string = 'main'
 param containerRegistryName string
+param secrets array = []
 param env array = []
 param external bool = true
 param imageName string
-param keyVaultName string = ''
-param managedIdentity bool = !empty(keyVaultName)
 param targetPort int = 80
 
+@description('User assigned identity name')
+param identityName string = ''
+
 @description('CPU cores allocated to a single container instance, e.g. 0.5')
 param containerCpuCoreCount string = '0.5'
 
 @description('Memory allocated to a single container instance, e.g. 1Gi')
 param containerMemory string = '1.0Gi'
 
+resource userIdentity 'Microsoft.ManagedIdentity/userAssignedIdentities@2023-01-31' existing = {
+  name: identityName
+}
+
+module containerRegistryAccess '../security/registry-access.bicep' = {
+  name: '${deployment().name}-registry-access'
+  params: {
+    containerRegistryName: containerRegistryName
+    principalId: userIdentity.properties.principalId
+  }
+}
+
 resource app 'Microsoft.App/containerApps@2022-03-01' = {
   name: name
   location: location
   tags: tags
-  identity: { type: managedIdentity ? 'SystemAssigned' : 'None' }
+  // It is critical that the identity is granted ACR pull access before the app is created
+  // otherwise the container app will throw a provision error
+  // This also forces us to use an user assigned managed identity since there would no way to 
+  // provide the system assigned identity with the ACR pull access before the app is created
+  dependsOn: [ containerRegistryAccess ]
+  identity: {
+    type: 'UserAssigned'
+    userAssignedIdentities: { '${userIdentity.id}': {} }
+  }
```

Comment:
User assigned identities are actually our only option here to use managed identity.  The identity being used must already have `AcrPull` role assigned to it before the container app is created otherwise the provision will hang and fail.

This means `SystemAssigned` identity is out of scope since it would be impossible to assign `AcrPull` to this identity since it is created at the same time that the container app is provisioned.